tests/hypothesis_strategies.py

```diff
@@ -0,0 +1,158 @@
+from hypothesis.extra.pandas import columns, data_frames
+
+from baybe.exceptions import NumericalUnderflowError
+from baybe.parameters.categorical import (
+    CategoricalEncoding,
+    CategoricalParameter,
+    TaskParameter,
+)
+from baybe.parameters.custom import CustomDiscreteParameter
+from baybe.parameters.numerical import (
+    NumericalContinuousParameter,
+    NumericalDiscreteParameter,
+)
+from baybe.parameters.substance import SubstanceEncoding, SubstanceParameter
+from baybe.utils import DTypeFloatNumpy
+
+_largest_lower_interval = np.nextafter(
+    np.nextafter(np.inf, 0, dtype=DTypeFloatNumpy), 0, dtype=DTypeFloatNumpy
+)
+"""
+The largest possible value for the lower end of a continuous interval such that there
+still exists a larger but finite number for the upper interval end.
+"""
+
+_decorrelate = st.one_of(
+    st.booleans(),
+    st.floats(min_value=0.0, max_value=1.0, exclude_min=True, exclude_max=True),
+)
+"""A strategy that generates decorrelation settings."""
+
+parameter_name = st.text(min_size=1)
+"""A strategy that generates parameter names."""
+
+categories = st.lists(st.text(min_size=1), min_size=2, unique=True)
+"""A strategy that generates parameter categories."""
+
+
+@st.composite
+def smiles(draw: st.DrawFn):
+    """Generate short SMILES strings."""
+    n_atoms = draw(st.integers(min_value=0, max_value=19))
+    string = "C"
+    for _ in range(n_atoms):
+        next_atom = random.choice("CNO") if string[-1] == "C" else random.choice("C")
+        string += next_atom
+    return string
+
+
+@st.composite
+def substance_data(draw: st.DrawFn):
+    """Generate data for class:`baybe.parameters.substance.SubstanceParameter`."""
+    names = draw(st.lists(st.text(min_size=1), min_size=2, max_size=10, unique=True))
+    substances = draw(st.lists(smiles(), min_size=len(names), max_size=len(names)))
+    return dict(zip(names, substances))
+
+
+@st.composite
+def custom_encodings(draw: st.DrawFn):
+    """Generate data for class:`baybe.parameters.custom.CustomDiscreteParameter`."""
+    index = st.lists(st.text(min_size=1), min_size=2, max_size=10, unique=True)
+    cols = columns(
+        names_or_number=10,
+        elements=st.floats(allow_nan=False, allow_infinity=False),
+        unique=True,
+        dtype=DTypeFloatNumpy,
+    )
+    return draw(data_frames(index=index, columns=cols))
+
+
+@st.composite
+def numerical_discrete_parameter(
+    draw: st.DrawFn,
+):
+    """Generate class:`baybe.parameters.numerical.NumericalDiscreteParameter`."""
+    name = draw(parameter_name)
+    values = draw(
+        st.lists(
+            st.one_of(
+                st.integers(),
+                st.floats(allow_infinity=False, allow_nan=False),
+            ),
+            min_size=2,
+            unique=True,
+        )
+    )
+
+    # Reject examples where the tolerance validator cannot be satisfied
+    try:
+        return NumericalDiscreteParameter(name=name, values=values)
```

Comment:
Yes and no, would appreciate your thoughts here:

So far, I deliberately did **not** add it, because providing a **valid** choice would require copying the same logic that is conducted **inside** the parameter class, introducing unnecessary redundancy and coupling. The current approach simply says "let the class itself decide what is appropriate", which is very robust and maintainable. Also, this type of rejection sampling is a core principle of hypothesis itself. 

A different reason why we might want to include it though (i.e. not only to pass validation) is code coverage. If we think that different tolerance values should be tested, then we should change the code. Here, we then run into the question if we want to get appropriate values by rejection sampling (current approach) or by replicating the acceptance-logic to draw valid values in the first place.

Reply:
im always pro code coverage 
what about creating a common utility that both the test and the implementation use to get the smallest/largest possible tolerance

Reply:
Done, the tolerance is now generated as well. Did not do it via utility though because in the parameter class, the necessary sorting step is already done in the attribute converter, so it cannot be extracted nicely into a utility. If you are happy with the code, please resolve 👍🏼